### PR TITLE
Feature/items

### PR DESCRIPTION
**Pull request by tschellenbach**
/**
 * It's common for messaging UIs to interleave and group messages
 *
 * - subsequent messages from the same user are typically grouped
 * - read state is typically shown (either as part of the message or separately)
 * - date separators are common
 * - typing indicators are typically shown at the bottom
 *
 * The MessageListItemLiveData class merges the livedata objects for messages, read state and typing
 *
 * It improves upon the previous Java code in the sample app in a few ways
 * - Kotlin
 * - Typing indicators can be turned on/off
 * - Date separators can be turned off and are configurable
 * - Read state matches to the right message
 * - Leverages MediatorLiveData to improve handling of null values
 * - Efficient algorithm for updating read state
 * - Efficient code for updating typing state
 * - Makes the MessageListItem immutable to prevent future bugs
 * - Improved test coverage
*/

Stats, note that typing and read events are more common than messages:

```
changing messages 100 times on a message list with 312 items took 31 milliseconds
changing typing information 100 times on a message list with 312 items took 2 milliseconds
changing read information 100 times on a message list with 312 items took 4 milliseconds
```

### CODE REVIEW

**Review by JcMinarro — CHANGES_REQUESTED** [review]
Why are we moving components from the UI library to the Offline Support SDK??

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    /**
+     * Typing changes are the most common changes on the message list
+     * Note how they don't recompute the message list, but only add to the end
+     */
+    internal fun typingChanged(users: List<User>): List<MessageListItem> {
+        val newTypingUsers = users.filter { it.id != currentUser.id }
+
+        return if (newTypingUsers != typingUsers) {
```
Comment (ogkuzmin) [inline livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:103]: You even can avoid storing current typingUsers. You can do the same in such way:
``` kotlin 
internal fun typingChanged(users: List<User>) {
        val newTypingUsers = users.filter { it.id != currentUser.id }
        val typingItems = usersAsTypingItems(newTypingUsers)
        val messageItemsWithReads = addReads(groupMessages(), readsLd.value)
        val newValue = wrapMessages(messageItemsWithReads + typingItems)
        if (value != newValue) {
            value = newValue
        } 
    }
```

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            typingUsers = newTypingUsers
+            typingItems = usersAsTypingItems()
+            val out = messageItemsWithReads + typingItems
+            value = wrapMessages(out)
```
Comment (ogkuzmin) [inline livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:107]: Is it 100 % defined, that such methods will be called only in main thread? I mean setting value to LiveData possible only from MainThread

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItem.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * MessageListItem is a sealed class with everything that is typically displayed in a message list
+ * - DateSeparatorItem
+ * - MessageItem
+ * - TypingItem
+ * - ThreadSeparatorItem
+ * - ReadStateItem
+ */
+sealed class MessageListItem {
```
[the remark this answers is not in the document]
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItem.kt:18]: the behaviour to group messages and combine read state, separators and message groups is universal to messaging apps regardless of their UI. imagine that you want to use Stream but build your own UI. It doesn't matter if you you're using compose, or the current layout system. You still want to have message groups, but you don't want to use our UI lib.

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+class MessageListItemLiveData(
+    val currentUser: User,
+    val messagesLd: LiveData<List<Message>>,
+    val readsLd: LiveData<List<ChannelUserRead>>,
+    val typingLd: LiveData<List<User>>? = null,
+    val isThread: Boolean = false,
+    val dateSeparator: ((previous: Message?, current: Message) -> Boolean)? = null
+) : MediatorLiveData<MessageListItemWrapper>() {
+
+    private var hasNewMessages: Boolean = false
```
[the remark this answers is not in the document]
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:58]: performance, you don't want to recompute everything when only the typing or read state changes.

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt (hunk 2 above)
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:107]: In our app we call it on the main thread, but indeed would be good to check

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        }
+    }
+
+    internal fun messagesChanged(messages: List<Message>): List<MessageListItem> {
+        messageItemsBase = groupMessages()
+        messageItemsWithReads = addReads(messageItemsBase, readsLd.value)
+        val out = messageItemsWithReads + typingItems
+        val wrapped = wrapMessages(out)
+        value = wrapped.copy(hasNewMessages = hasNewMessages)
+        return out
```
[the remark this answers is not in the document]
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:86]: it makes it easier/more isolated to test

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt (hunk 2 above)
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:107]: Will decorate with https://developer.android.com/reference/kotlin/androidx/annotation/UiThread

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ogkuzmin) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:58]: You would be surprise if you started to measure everything from request to API to showing in UI, todays devices are so powerful that calculations doesn't impact performance. The most impacting things are threading and IO operations. I could investigate and measure our app and decide what needs to be speeded up

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ogkuzmin) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:86]: In common it's no good when tests force you to change design. You can check same with other options

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+ *   return if (previous==null) {
+ *       true
+ *   } else {
+ *       (message.getCreatedAtOrThrow().time - previous.getCreatedAtOrThrow().time) > (60 * 60 * 3)
+ *   }
+ * }
+ *
+ */
+class MessageListItemLiveData(
+    val currentUser: User,
```
[the remark this answers is not in the document]
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:50]: +1

File: livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        }
+
+        return messagesCopy
+    }
+
+    private fun wrapMessages(items: List<MessageListItem>): MessageListItemWrapper {
+        return MessageListItemWrapper(items = items, isThread = isThread, isTyping = typingUsers.isNotEmpty(), hasNewMessages = hasNewMessages)
+    }
+
+    private fun usersAsTypingItems(): List<MessageListItem> {
```
[the remark this answers is not in the document]
Comment (tschellenbach) [reply livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt:216]: agreed

**Comment by tschellenbach** [comment]
Will move this to the UI layer instead. The concept of grouping messages is independent from the UI. But we can revisit that topic in the future.

**Pull request closed without merging**